### PR TITLE
backup: custom S3 endpoint support

### PR DESCRIPTION
**Pull request by givia**
this enables backup on other implementations of S3, such as ceph and minio.

### CODE REVIEW

**Comment by hongchaodeng** [comment]
@givia Can you create an issue and describe your use case?

**Comment by givia** [comment]
Fixes #1720

**Comment by xiang90** [comment]
Have you tried this manually. is it possible to get this test covered somehow? (setting up a minion server for example?)

**Comment by givia** [comment]
Good idea. I will add the tests.

**Comment by givia** [comment]
@xiang90 @hongchaodeng

File: pkg/apis/etcd/v1beta2/backup.go [hunk 1]
```diff
@@ -42,4 +42,7 @@ type S3Source struct {
 	//
 	// AWSSecret overwrites the default etcd operator wide AWS credential and config.
 	AWSSecret string `json:"awsSecret,omitempty"`
+
+	// AWSEndpoint
+	AWSEndpoint string `json:"awsEndpoint,omitempty"`
```
Comment (xiang90) [inline pkg/apis/etcd/v1beta2/backup.go]: this should be just endpoint? we also need to document this better.

empty means aws s3, which is by default. you can fill in endpoints if you wish to use minion or some other compatible provider?

File: pkg/apis/etcd/v1beta2/restore_types.go [hunk 2]
```diff
@@ -65,6 +65,8 @@ type S3RestoreSource struct {
 	//
 	// AWSSecret overwrites the default etcd operator wide AWS credential and config.
 	AWSSecret string `json:"awsSecret"`
+
+	AWSEndpoint string `json:"awsEndpoint"`
```
Comment (xiang90) [inline pkg/apis/etcd/v1beta2/restore_types.go]: documentation is needed

**Comment by xiang90** [comment]
@alexandrem  can you please review this PR?

**Comment by alexandrem** [comment]
Sorry I was away for many days.

This branch required merging work from master because things had diverged a lot.

I have an updated version in alexandrem:aws-endpoint

Not quite sure though, because I can't get the backup operation to work.

Using this:

```apiVersion: "etcd.database.coreos.com/v1beta2"
kind: "EtcdBackup"
metadata:
  name: etcd-cluster-backup
  namespace: test
spec:
  etcdEndpoints:
  - http://etcd-cluster-client:2379
  storageType: S3
  s3:
    path: test/givia-s3
    awsSecret: aws
    awsEndpoint: http://minio-minio-svc.default.svc.cluster.local:9000/```

But it fails at the etcd cluster client connection:

```time="2018-01-12T21:18:05Z" level=error msg="error syncing etcd backup (test/etcd-cluster-backup): failed to save snapshot (create etcd client failed: no running etcd pods found)" pkg=controller 
```

Even though I have a fresh cluster running.

```$ kubectl get etcdcluster -n test etcd-cluster -o yaml
apiVersion: etcd.database.coreos.com/v1beta2
kind: EtcdCluster
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"etcd.database.coreos.com/v1beta2","kind":"EtcdCluster","metadata":{"annotations":{},"labels":{"app":"etcd"},"name":"etcd-cluster","namespace":"test"},"spec":{"size":3,"version":"3.2.10"}}
  clusterName: ""
  creationTimestamp: 2018-01-12T21:17:19Z
  generation: 0
  labels:
    app: etcd
  name: etcd-cluster
  namespace: test
  resourceVersion: "11829163"
  selfLink: /apis/etcd.database.coreos.com/v1beta2/namespaces/test/etcdclusters/etcd-cluster
  uid: f9313644-f7dd-11e7-abb1-fa163e9b6160
spec:
  baseImage: gcr.io/etcd-development/etcd
  size: 3
  version: 3.2.10
status:
  clientPort: 2379
  conditions:
  - lastTransitionTime: 2018-01-12T21:18:00Z
    lastUpdateTime: 2018-01-12T21:18:00Z
    reason: Cluster available
    status: "True"
    type: Available
  currentVersion: 3.2.10
  members:
    ready:
    - etcd-cluster-0000
    - etcd-cluster-0001
    - etcd-cluster-0002
  phase: Running
  serviceName: etcd-cluster-client
  size: 3
  targetVersion: ""```

**Comment by xiang90** [comment]
close in favor of close in favor of #1855.

/cc @alexandrem

**Pull request closed without merging**